File: backend/app/file/service/file/file.go
package file

type Chunk struct {
	ChunkIndex int64 `bson:"ChunkIndex"`
	ChunkOK    bool  `bson:"ChunkOK"`
}

type File struct {
	Hash         string  `bson:"Hash"`
	Status       int32   `bson:"Status"`
	ChunkNums    int64   `bson:"ChunkNums"`
	CurrentIndex int64   `bson:"CurrentIndex"`
	SegmentSize  int64   `bson:"SegmentSize"`
	Bucket       string  `bson:"Bucket"`
	FilePath     string  `bson:"FilePath"`
	Chunks       []Chunk `bson:"Chunks"`
}

File: frontend/src/apis/file.ts
import {get, Instance, post} from "./http";

export interface Config {
  configID: string
  configType: string
  version: string
  configData: any
}

export const startUploadFile = (hash: string, fileName: string, chunkNums: number, segmentSize: number) => {
  return post<Config>(`/files/start`, {
    hash,
    fileName,
    chunkNums,
    segmentSize
  })
}

export const uploadChunk = (chunk: Blob, hash: string, chunkIndex: number) => {
  const formData = new FormData()
  formData.set("chunk", chunk)
  formData.set("hash", hash)
  formData.set("chunkIndex", chunkIndex + "")
  return Instance.postForm("/files/chunk", formData)
}

interface FileInfo {
  hash: string;
  status: 0 | 1 | 2
  current_index: number
}

export const queryFile = (hash: string) => {
  return get<FileInfo>(`/files/${hash}`)
}

export const queryFileURL = (hash: string) => {
  return get<{ url: string }>(`/files/${hash}/public_url`)
}